### ISSUE TITLE
Reproduction results

### ISSUE THREAD

**Issue by wulin97** [DylanOrange/geal#9]
Hi, thanks for releasing the code and checkpoint!

I re-trained the models using the code (stage 1 and 2). The performance is obviously lower than the official metrics reported in the paper and the checkpoint provided.

Could you please confirm whether the results in the paper were obtained using exactly the same code and hyperparameters as the ones released here? or suggest other modifications I should make to reproduce the result?

Results Comparison:

Metric | Official ckpt | My reproduction
IOU	0.225	0.239
AUC	0.868	0.844
SIM	0.632	0.512
MAE	0.096	0.132

**Comment by huamo555** [comment]
> Hi, thanks for releasing the code and checkpoint!
> 
> I re-trained the models using the code (stage 1 and 2). The performance is obviously lower than the official metrics reported in the paper and the checkpoint provided.
> 
> Could you please confirm whether the results in the paper were obtained using exactly the same code and hyperparameters as the ones released here? or suggest other modifications I should make to reproduce the result?
> 
> Results Comparison:
> 
> Metric | Official ckpt | My reproduction IOU 0.225 0.239 AUC 0.868 0.844 SIM 0.632 0.512 MAE 0.096 0.132

Did you successfully resolve the issue? I'm currently getting an error when running stage2.

<img width="1054" height="273" alt="Image" src="https://github.com/user-attachments/assets/b4d144b7-d596-411c-9541-add314b3859a" />

**Comment by LemonQC** [comment]
I am also unable to reproduce the results. Additionally, I found that the author's paper mentions a batch size of 12, but it can only be modified to 12 in the first stage and 8 in the second stage. After being modified to 12, an error occurs.

**Comment by DylanOrange** [comment]
Hi,

The two-stage pipeline is sensitive, and in practice, I also need some parameter tuning. Here are a few suggestions based on my experience:

For Stage 1, similar to OOAL ([CVPR'24](https://github.com/Reagan1311/OOAL)), adding simple 2D evaluation or metrics (e.g., visualization checks) helps ensure the 2D prediction network is producing meaningful outputs before entering Stage 2.

For Stage 1, you can also try this [function](https://github.com/DylanOrange/geal/blob/3048bb8c859b04b3d7569e2bc186f9880c7109ab/renderer/render_utils.py#L69) to project the 2D affordance back to the 3D point cloud directly to visualize the results, using the stored Gaussian contribution ID.

Handle the foreground–background imbalance. In stage 1, the positive region is typically much smaller than the negative region. Separating positive and negative pixels when computing the BCE loss (or using re-weighted BCE / focal loss) may help a lot.

Regarding the batch size, I personally didn’t encounter the issue where changing the batch size causes the training to fail. You can share the exact error message when setting the second-stage batch size to 12.


> I am also unable to reproduce the results. Additionally, I found that the author's paper mentions a batch size of 12, but it can only be modified to 12 in the first stage and 8 in the second stage. After being modified to 12, an error occurs.

**Comment by LemonQC** [comment]
Thanks， is this setting right？

<img width="869" height="954" alt="Image" src="https://github.com/user-attachments/assets/26802dbe-273a-4321-9e00-86104e32ab1a" />


> Hi,
> 
> The two-stage pipeline is sensitive, and in practice, I also need some parameter tuning. Here are a few suggestions based on my experience:
> 
> For Stage 1, similar to OOAL ([CVPR'24](https://github.com/Reagan1311/OOAL)), adding simple 2D evaluation or metrics (e.g., visualization checks) helps ensure the 2D prediction network is producing meaningful outputs before entering Stage 2.
> 
> For Stage 1, you can also try this [function](https://github.com/DylanOrange/geal/blob/3048bb8c859b04b3d7569e2bc186f9880c7109ab/renderer/render_utils.py#L69) to project the 2D affordance back to the 3D point cloud directly to visualize the results, using the stored Gaussian contribution ID.
> 
> Handle the foreground–background imbalance. In stage 1, the positive region is typically much smaller than the negative region. Separating positive and negative pixels when computing the BCE loss (or using re-weighted BCE / focal loss) may help a lot.
> 
> Regarding the batch size, I personally didn’t encounter the issue where changing the batch size causes the training to fail. You can share the exact error message when setting the second-stage batch size to 12.
> 
> > I am also unable to reproduce the results. Additionally, I found that the author's paper mentions a batch size of 12, but it can only be modified to 12 in the first stage and 8 in the second stage. After being modified to 12, an error occurs.

**Comment by LemonQC** [comment]
Error occurs when stage 2， the batch size is set to 12：

<img width="2997" height="361" alt="Image" src="https://github.com/user-attachments/assets/63ed108b-8036-404c-ae9f-14550ae397b5" />

**Comment by DylanOrange** [comment]
this looks like a NumPy issue caused by trying to convert a list of elements with inconsistent shapes into an array. In practice, this often happens when predictions/labels are stored per batch, and the last batch has a different size, leading to an inhomogeneous list when calling np.array(...).

> Error occurs when stage 2， the batch size is set to 12：
> 
> <img alt="Image" width="2000" height="361" src="https://private-user-images.githubusercontent.com/30914380/537922587-63ed108b-8036-404c-ae9f-14550ae397b5.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njg5MDAyODUsIm5iZiI6MTc2ODg5OTk4NSwicGF0aCI6Ii8zMDkxNDM4MC81Mzc5MjI1ODctNjNlZDEwOGItODAzNi00MDRjLWFlOWYtMTQ1NTBhZTM5N2I1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTIwVDA5MDYyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZjMWFkMjU0NjhmZjZhY2RlZTBlOTg2OTJmOWYwMzAzNjJhMTJiMjgzYTI3YTQ5Y2M2OGIxMzFkNzk0MzE5OTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.SgaIjeXIFk989TqWHoV_kQh1F5EnM8-tyKyKNX3K6DA">

**Comment by LemonQC** [comment]
I have solved this problem by the following setting:

<img width="556" height="191" alt="Image" src="https://github.com/user-attachments/assets/7c91569e-6361-4a86-8727-05de06f72f4d" />
,I wanna ask in stage2.yaml, should i set pretrained_2d and checkpoint_path as follow

<img width="1368" height="466" alt="Image" src="https://github.com/user-attachments/assets/69a19ecc-8b3e-4adb-b2a8-d4594f6b60d7" />

**Comment by DylanOrange** [comment]
> I have solved this problem by the following setting:
> 
> <img alt="Image" width="556" height="191" src="https://private-user-images.githubusercontent.com/30914380/537934738-7c91569e-6361-4a86-8727-05de06f72f4d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njg5MDEyMDgsIm5iZiI6MTc2ODkwMDkwOCwicGF0aCI6Ii8zMDkxNDM4MC81Mzc5MzQ3MzgtN2M5MTU2OWUtNjM2MS00YTg2LTg3MjctMDVkZTA2ZjcyZjRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTIwVDA5MjE0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVjNjkzYzZjYTk4OGRmYjlmNWM4MzUyZWMzODE3OWI5NWQ5YjUxOWUwZDM5MWEwNTU4MDE1NzI3ZjFmMmE3YzcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ugNYtYfiCzXlMezmWuNBoTOMSRkiPx8i1k9Bdw04cr8"> ,I wanna ask in stage2.yaml, should i set pretrained_2d and checkpoint_path as follow <img alt="Image" width="1368" height="466" src="https://private-user-images.githubusercontent.com/30914380/537935126-69a19ecc-8b3e-4adb-b2a8-d4594f6b60d7.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njg5MDEyMDgsIm5iZiI6MTc2ODkwMDkwOCwicGF0aCI6Ii8zMDkxNDM4MC81Mzc5MzUxMjYtNjlhMTllY2MtOGIzZS00YWRiLWIyYTgtZDQ1OTRmNmI2MGQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTIwVDA5MjE0OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNkYTBkM2Y4ZGJiMzBmMDk4YTlhODk5ODg4MjI0N2JlZGVkYzI5YWIyZjBmMWEzNjQ5NzExYmM5Y2RkNjJjOWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.1-wyrZ0k7ICpMjMkdjLODNHPw4QWFDJKb02oE40TPOM">

Yeah

**Comment by LemonQC** [comment]
Thanks a lot. I will try my best to reproduce the results.

**Comment by huamo555** [comment]
> I have solved this problem by the following setting:
> 
> <img alt="Image" width="556" height="191" src="https://private-user-images.githubusercontent.com/30914380/537934738-7c91569e-6361-4a86-8727-05de06f72f4d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njk1OTI2MTQsIm5iZiI6MTc2OTU5MjMxNCwicGF0aCI6Ii8zMDkxNDM4MC81Mzc5MzQ3MzgtN2M5MTU2OWUtNjM2MS00YTg2LTg3MjctMDVkZTA2ZjcyZjRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTI4VDA5MjUxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFmMjZjZDY2ZWU4NzQ4ZGE2MjYxN2U3NjZlZjE1MjU0MTFkM2JhMGU1NGRlODM1NzI4OWNlOWExMDZlMWQ3NjQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.SrcLAARwBwqvCJerCHwcV68CZFM0etTpOp-L9A1moVM"> ,I wanna ask in stage2.yaml, should i set pretrained_2d and checkpoint_path as follow <img alt="Image" width="1368" height="466" src="https://private-user-images.githubusercontent.com/30914380/537935126-69a19ecc-8b3e-4adb-b2a8-d4594f6b60d7.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njk1OTI2MTQsIm5iZiI6MTc2OTU5MjMxNCwicGF0aCI6Ii8zMDkxNDM4MC81Mzc5MzUxMjYtNjlhMTllY2MtOGIzZS00YWRiLWIyYTgtZDQ1OTRmNmI2MGQ3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTI4VDA5MjUxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWExZGJhNDkwMWRjNmEzZGQyODU2MzUwOWI1MDMyYmU5OWQxYTdmOGRkNzU3YzRmY2U0NzVmYzFjYmMwMmExNzYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.buvGE6JcgVOfAKeJsy42mwd2SubfTNapmK1EL0F0yDk">

Have you successfully reproduced the result?

**Comment by LemonQC** [comment]
No, the results are still different

**Comment by huamo555** [comment]
> 不，结果还是不同

Hello, what is the current result you have reproduced?

**Comment by LemonQC** [comment]
<img width="1620" height="166" alt="Image" src="https://github.com/user-attachments/assets/3755da44-2c72-450d-b1be-11f0e614701c" />

**Comment by huamo555** [comment]
> <img alt="Image" width="1620" height="166" src="https://private-user-images.githubusercontent.com/30914380/541570912-3755da44-2c72-450d-b1be-11f0e614701c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njk3NTQ5NjYsIm5iZiI6MTc2OTc1NDY2NiwicGF0aCI6Ii8zMDkxNDM4MC81NDE1NzA5MTItMzc1NWRhNDQtMmM3Mi00NTBkLWIxYmUtMTFmMGU2MTQ3MDFjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAxMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMTMwVDA2MzEwNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIwMmQxZDg0NjVkNDZhNmIwZWI5ZTQ1M2NkM2Y1MjAzODA3MmE2NWIyYzUyODc5MWMyODQyNDcyY2M3NDdjMDYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.yNrSsEztGSGnC8e3HPGBUQMF2pBfhkBhklGhGWdx4TI">

My results are similar to yours, haha, so it's probably not a reproduction issue.

<img width="523" height="36" alt="Image" src="https://github.com/user-attachments/assets/1405cfc4-7c88-42dd-b481-4002667d251b" />

**Comment by DylanOrange** [comment]
Hi, thanks for sharing the metrics and the observation.

I agree it’s a bit counter-intuitive at first: if IoU/AUC are already very strong, it usually means the geometry/localization is working well. When SIM/MAE are still lagging, in my experience it often points to the value regression / distribution matching part being under-constrained, and those metrics can be quite sensitive to the exact loss formulation and weighting.

One practical thing you could try (optional) is adding a small amount of direct supervision that aligns with the SIM/MAE objective on the 3D branch (e.g., an MAE term, or a SIM-oriented loss), and then tuning its weight gently. This often can help lift SIM/MAE while keeping localization largely unchanged, although there may be a small trade-off depending on the dataset and training setup. From my experience, this is more of an engineering knob, helpful if you’re optimizing specifically for MAE/SIM, but not necessary if your goal is a strict paper-style reproduction.
